c++/src/kj/compat/http.c++

```diff
@@ -4818,6 +4897,17 @@ private:
           auto& headers = httpInput.getHeaders();
 
           currentMethod = request.method;
+
+          SuspendableRequest suspendable(*this, request.method, request.url, headers);
+          auto maybeService = factory(suspendable);
+
+          if (suspended) {
+            return httpOutput.flush().then([this]() { return false; });
```

Comment:
Probably so -- I didn't completely think this line through. I'll verify it's safe to remove.

Reply:
https://github.com/capnproto/capnproto/pull/1231/commits/476e3555403d1f554aa49aa4eb6f976b0004811d

I'm convinced we don't need to flush the output, because we only call `loop()` on a new Connection, or after a call to `httpOutput.flush()`. But there are a couple calls to `httpOutput.flush()` above this location, specifically where it checks `if (timedOut)` and `if (closed)`, where I'd also expect this logic to apply. Perhaps those calls are extraneous, too?

Reply:
Hmm good question.

Actually, maybe this does matter in the case of pipelined HTTP requests, i.e. where a new request arrives before the previous response finished. That's technically legal, though rarely used due to head-of-line blocking problems. I suppose the `flush()` here ensures that the _previous_ response is done sending? In that case it might actually be important when suspending. (In our specific use case, it won't be relevant, since we know we have a proxy in front of us that won't do pipelining, but still good to get it right here.)

I guess this also awkwardly means that we have to tell users of the new interface that they must not actually transfer the connection to a new thread until `listenHttpCleanDrain()` has actually returned in the original thread, even though they were given the SuspendedRequest before that. Makes me wonder if we should actually return the SuspendedRequest...

Reply:
> I suppose the flush() here ensures that the previous response is done sending? In that case it might actually be important when suspending.

I don't think it does. It looks like `flush()` just waits on the write queue, not for `HttpOutputStream::inBody` to equal false or any particular state change in the output stream. Nevertheless, the only way I can see for `HttpOutputStream::inBody` to be true when we read the next request's headers is for an `Http*EntityWriter` returned from `Response::send()` to outlive the promise returned from `service.request()`. I would imagine that should be considered a programming error, though.

Reply:
Hmm are you saying `HttpServer` doesn't deliver pipelined requests to the app concurrently in the first place? I think the original version did but I suppose I might have removed that support to simplify things at some point.

Reply:
> Hmm are you saying HttpServer doesn't deliver pipelined requests to the app concurrently in the first place?

I looked again, and yeah, I don't see how pipelined requests could be delivered concurrently. The connection loop waits on the HttpService::request() promise (https://github.com/capnproto/capnproto/blob/master/c%2B%2B/src/kj/compat/http.c%2B%2B#L4838-L4840), after which point I think we can assume the lifetime of any response Http*EntityWriter body should have ended (and called `finishBody()` / `abortBody()` as needed). Then, before recursing, it the loop flushes the output: https://github.com/capnproto/capnproto/blob/master/c%2B%2B/src/kj/compat/http.c%2B%2B#L4876-L4883

So, I think we have a clean output stream at the point where we decide whether to suspend, and the other calls to `flush()` in the beginning of the loop are leftover from when the loop supported request pipelining.

We could consider returning a SuspendedRequest anyway, to keep the option of reimplementing request pipelining open.

Reply:
> We could consider returning a SuspendedRequest anyway, to keep the option of reimplementing request pipelining open.

I seem to recall thinking it felt cleaner that way, too, but I don't remember my exact reasoning. I guess I leave it up to you to decide whether it's worth doing.